### PR TITLE
Ensure DataProtection can be used in trimmed apps.

### LINKED ISSUE #47695: Respond to IL2026 warnings in DataProtection originating from EncryptedXml
With https://github.com/dotnet/runtime/pull/84468, `EncryptedXml` is now marked as `RequiresUnreferencedCode`. This is because

```
The algorithm implementations referenced in the XML payload might be removed. Ensure the required algorithm implementations are preserved in your application.
```

From discussion in https://github.com/dotnet/aspnetcore/issues/47410#issuecomment-1497423603, we should:

1. Figure out the list of commonly used crypto algorithms used by DataProtection, and add `[DynamicDependency]` attributes for them.
2. We also add some trimming/aot tests to ensure DataProtection works correctly when trimmed.
3. DataProtection then suppresses the warnings coming from `System.Security.Cryptography.Xml`

### PR DESCRIPTION
**Pull request by eerhardt**
- Add a DynamicDependency to ensure Aes decryption works in EncryptedXmlDecryptor
- Suppress the warnings from EncryptedXml
- Add trimming tests to ensure these scenarios work correctly.
- Remove the RequiresUnreferencedCode attribute on AddAuthentication, since this is the only thing in that method that has warnings.

Fix https://github.com/dotnet/aspnetcore/issues/47695

### CODE REVIEW

File: src/Shared/TrimmingAttributes.cs [hunk 1]
```diff
@@ -154,6 +154,153 @@ public UnconditionalSuppressMessageAttribute(string category, string checkId)
...
+/// States a dependency that one member has on another.
+/// </summary>
+/// <remarks>
+/// This can be used to inform tooling of a dependency that is otherwise not evident purely from
+/// metadata and IL, for example a member relied on via reflection.
+/// </remarks>
+[AttributeUsage(
+    AttributeTargets.Constructor | AttributeTargets.Field | AttributeTargets.Method,
+    AllowMultiple = true, Inherited = false)]
+internal sealed class DynamicDependencyAttribute : Attribute
```
Comment (mitchdenny) [inline src/Shared/TrimmingAttributes.cs:167]: Why do we need this via shared source?

**Review by mitchdenny — COMMENTED** [review]
More of a question than a blocking issue. Seems we are copying a type definition that already exists in code analysis?

File: src/DataProtection/DataProtection/test/Microsoft.AspNetCore.DataProtection.TrimmingTests/Constants.cs [hunk 2]
```diff
@@ -0,0 +1,93 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+internal static class Constants
+{
+    public const string Password = "password";
```
Comment (danmoseley) [inline src/DataProtection/DataProtection/test/Microsoft.AspNetCore.DataProtection.TrimmingTests/Constants.cs:6]: Credscan is going to flag anything that looks like a password/token/private key. I think it's fine to check in am suppress later, but fyi "placeholder" will not trigger it, while "password" will...

File: src/Shared/TrimmingAttributes.cs (hunk 1 above)
Comment (eerhardt) [reply src/Shared/TrimmingAttributes.cs:167]: These attributes were only introduced in `net5.0`. You can see what versions they are supported in https://apisof.net/catalog/5e049fae-038e-f176-2964-3db4b6b04ad5. DataProtection builds for `netstandard2.0` and `net4xx` where these attributes aren't defined. So when building for those older TFMs, we copy the attributes into the assembly so we don't need to use a bunch of `#if` defs.

File: src/DataProtection/DataProtection/test/Microsoft.AspNetCore.DataProtection.TrimmingTests/Constants.cs (hunk 2 above)
Comment (eerhardt) [reply src/DataProtection/DataProtection/test/Microsoft.AspNetCore.DataProtection.TrimmingTests/Constants.cs:6]: These were copied from

https://github.com/dotnet/runtime/blob/4679e59ac4431380ff2455b471c9b3e9fda503e6/src/libraries/System.Net.Http.WinHttpHandler/tests/UnitTests/ClientCertificateHelper.cs#L176-L223

That code has a commented out SuppressMessage - not sure if that is doing anything.

If we don't use "password", we will need a new certificate.

File: src/Shared/TrimmingAttributes.cs (hunk 1 above)
Comment (mitchdenny) [reply src/Shared/TrimmingAttributes.cs:167]: TIL :)

File: src/DataProtection/DataProtection/test/Microsoft.AspNetCore.DataProtection.TrimmingTests/Constants.cs (hunk 2 above)
Comment (danmoseley) [reply src/DataProtection/DataProtection/test/Microsoft.AspNetCore.DataProtection.TrimmingTests/Constants.cs:6]: eh, fine for now

Comment (eerhardt) [reply src/DataProtection/DataProtection/test/Microsoft.AspNetCore.DataProtection.TrimmingTests/Constants.cs:6]: I added a suppression for the cert, like in the above link. Not sure if I need one for the password. How often does Credscan run on our code?

**Review by wtgodbe — APPROVED** [review]
Build bits seem reasonable

File: src/DataProtection/DataProtection/src/XmlEncryption/EncryptedXmlDecryptor.cs [hunk 3]
```diff
@@ -44,6 +45,11 @@ public EncryptedXmlDecryptor(IServiceProvider? services)
     /// </summary>
     /// <param name="encryptedElement">An encrypted XML element.</param>
     /// <returns>The decrypted form of <paramref name="encryptedElement"/>.</returns>
+#pragma warning disable SYSLIB0022 // Rijndael types are obsolete
+    [DynamicDependency(DynamicallyAccessedMemberTypes.PublicParameterlessConstructor, typeof(RijndaelManaged))]
```
Comment (amcasey) [inline src/DataProtection/DataProtection/src/XmlEncryption/EncryptedXmlDecryptor.cs:49]: Is this actually used or does it just get instantiated and ignored?

**Review by amcasey — APPROVED** [review]
I have limited knowledge of this code, but the change seems low-risk to me.  Nits below.

File: src/DataProtection/DataProtection/src/XmlEncryption/EncryptedXmlDecryptor.cs (hunk 3 above)
Comment (amcasey) [reply src/DataProtection/DataProtection/src/XmlEncryption/EncryptedXmlDecryptor.cs:49]: Per our offline discussion, it's used and important.  I might say something like "Rijndael (aka AES) is used in nearly all scenarios.  If we find another important family, we should add it here as well.  In the meantime, a useful exception will be thrown."

**Review by amcasey — APPROVED** [review]
LGTM.  Thanks for the offline explanation!

File: src/DataProtection/DataProtection/src/XmlEncryption/CertificateXmlEncryptor.cs [hunk 4]
```diff
@@ -3,6 +3,7 @@
 
 using System;
 using System.Diagnostics.CodeAnalysis;
+using System.Security.Cryptography;
```
[the remark this answers is not in the document]
Comment (eerhardt) [reply src/DataProtection/DataProtection/src/XmlEncryption/CertificateXmlEncryptor.cs:6]: Nope. Removed.

File: src/DataProtection/DataProtection/src/XmlEncryption/CertificateXmlEncryptor.cs [hunk 5]
```diff
@@ -76,6 +77,8 @@ public EncryptedXmlInfo Encrypt(XElement plaintextElement)
         return new EncryptedXmlInfo(encryptedElement, typeof(EncryptedXmlDecryptor));
     }
 
+    [UnconditionalSuppressMessage("AOT", "IL2026:RequiresUnreferencedCode",
```
[the remark this answers is not in the document]
Comment (eerhardt) [reply src/DataProtection/DataProtection/src/XmlEncryption/CertificateXmlEncryptor.cs:80]: Done.

File: src/DataProtection/DataProtection/src/XmlEncryption/EncryptedXmlDecryptor.cs (hunk 3 above)
Comment (eerhardt) [reply src/DataProtection/DataProtection/src/XmlEncryption/EncryptedXmlDecryptor.cs:49]: Added a comment.

**Comment by JamesNK** [comment]
Nice!

Does this mean we can move authentication/authorization to supported?

**Comment by eerhardt** [comment]
> Does this mean we can move authentication/authorization to the AOT supported column?

Some things still don't work. For example, the JWT bearer handler depends on System.IdentityModel.Tokens.Jwt, which is not AOT compatible.

For the other authentication modes, are there other 3rd party dependencies? Or are they all contained in this repo?

**Pull request merged**